docs/services.md

```diff
@@ -0,0 +1,84 @@
+# Services
+
+## Abstract
+
+In sct we have repeated pattern when we run code in parallel and time to time need to sync with it, get status, make it stop.
+This pattern is to be called service.
+List key features that distinguish service pattern:
+- Code is run in parallel
+- Code is doing repetitive task
+- You do one of the following:
+    - Check the code status
+    - Make it stop by command
+    - Want it to stop automatically when:
+        - Test is finished
+        - Node is destroyed
+        - Cluster is destroyed
+
+## Goals
+### Effortless lifecycle
+Make services stopped/killed automatically, with no additional code
+### Effortless dependency tracking
+Make services to automatically know their dependencies, so that they could be stopped/killed in proper manner, to avoid unexpected errors to pop up
+### Logging
+Make logging structured so that you could see which service generated an error and which node/cluster associated with it
+### Provide same interface for all threads/processes across sct
+Big problem now when we have tens of threads/processes managed in different manner. It makes hard to do things like:
+- Track them
+- Stop or kill them, especially to do it gracefully
+
+## Type of services
+
+* Types by attachment
+    * Detached Service
+       > Service that is not attached to node or cluster
+Life cycle of this type of service ends when test is finished
+It will be stopped/killed only after all it's dependants are stopped/killed
+
+    * Node Service
```

Comment:
for example a `c-s stress thread` can be such thing ?
also what about a docker running on a node ? i.e. it's a common think I would like to have for the stress commands.

Reply:
stress threads are fit perfectly, dockers running not sure, unless we come up with DockerService

Reply:
check out `DockerBasedStressThread`, which is the basis of what I'm referring to. it's a ugly hack to get us in direction of doing a docket base stress tool, and I'm using that for YCSB, Ndbench, and @aleksbykov used it for one of the cdc related tools.
I would love to do the same also for c-s, since the whole version control we do for it in the loader AMI is a complete disaster.

Also using dockers, enable to test lots of the implementation locally with integration tests, using one real scylla docker, and a stress command class

See those for examples:
https://github.com/scylladb/scylla-cluster-tests/blob/master/unit_tests/test_ycsb_thread.py
https://github.com/scylladb/scylla-cluster-tests/blob/master/unit_tests/test_ndbench_thread.py

(since they are a bit longer tests the the usual unittests, and resources housekeeping of those wasn't that good, we are not running them on PRs anymore)